### ISSUE TITLE
hi,mikerabat

### ISSUE THREAD

**Issue by PassByYou888** [mikerabat/mrmath#6]
hi,mikerabat

I plan to use mrmath (aiot / nb-iot) in Jetson for my project.
Assembly / AVX / SSE code cannot support armv7 / arm64 / MIPs
Please keep Pascal source code. thank you!

Best Regards
by.qq600585

**Comment by mikerabat** [comment]
Dear User,



On Thu, Dec 23, 2021 at 4:52 PM PassByYou888 ***@***.***>
wrote:

> hi,mikerabat
>
> I plan to use mrmath (aiot / nb-iot) in Jetson for my project.
> Assembly / AVX / SSE code cannot support armv7 / arm64 / MIPs
> Please keep Pascal source code. thank you!
>

Yeah... unfortunately due to lack of a platform to develop on I don't have
an ARM optimized
version.
With some work one could eventually only use the pascal files which at
least have some
optimizations in it (e.g. the blocked matrix multiplication is way faster
than the naive method).
Feel free to add the defines where needed ;)

kind regards
   Mike

**Comment by PassByYou888** [comment]
hi,dear mike,

Optimization belongs to Dever work all over the world

Paid engineers are responsible for hardware optimization, including Intel, MKL, daal, NVIDIA, of more, They made open source PR and got money. We are Pascal fans, so we try to use FPC to solve matrix computing

These engineers will also be responsible for optimization when needed: nlopt, eigen3, of more

im feel that FPC is not good at optimizing programming. I suggest giving a native source on the condition that it loses the feeling of cool oh mike,im so fun.

best regards
by. qq600585

**Comment by neurolabusc** [comment]
@PassByYou888 I agree that FPC is often not as fast as C code for numerical routines, and there are fewer optimized Pascal numerical libraries (which makes mrmath so important to our community). Over time, FPC has been steadily improving for x86-64. On the other hand, in my experience the FPC ARM code is robust but not very fast. Since the topic of this issue is ARM code, I suggest you try using [LLVM](https://wiki.freepascal.org/LLVM) to compile your Pascal code. For my M1 Mac, I found FPC-LLVM was [much faster](https://github.com/neurolabusc/AppleSiliconForNeuroimaging) than pure FPC.

If you want to see FPC numerical performance improve, I encourage you to join me in supporting [J. Gareth Moreton](https://www.patreon.com/curiouskit/posts) on Patreon. Over the last year he has been responsible for many of the peephole optimizations in FPC.

**Comment by xiaodongzhan** [comment]
hi,Mike:
Nice job!
**Please support Free Pascal!**
Thanks you!

**Comment by mikerabat** [comment]
Please note that the functionality is 100% pure pascal code (e.g. if you call the
InitMathFunctions(itFPU) then everything is mapped to pascal code...)
What is missing are the ifdefs for ARM targets. 
As I said you are free to put them in there since I don't have such a CPU just lying aroun....

**Comment by neurolabusc** [comment]
@PassByYou888 you might [look at this kludge](https://github.com/neurolabusc/FPCintrinsics).

**Comment by PassByYou888** [comment]
@neurolabusc very good
i'm in China. we can't use money to like  " J. Gareth Moreton". shit bitcoin to close my money for used.
im think,if you have the conditions to solve Pascal algorithm problems, you should go all out to contribute to Pascal future

maybe our can construct non liner matrix support for Pascal?

best regards
by. qq600585

**Comment by mikerabat** [comment]
Finally - I got help to create a version of the lib without the assembler routines - just pure Pascal.
The latest checkin actually introduces a new automatic compiler define
MRMATH_NOASM
based on the CPU target. Noasm is defined if no x86 nor x64 cpu target is used.

I also updated the packages.
Note that the unit tests were too complicated to fix since 99% are indeed asm tests
